src/app/cluster-building-blocks/QuieterReporting.h

```diff
@@ -0,0 +1,249 @@
+        const Nullable<T> & newValue;
+    };
+
+    using SufficientChangePredicate = std::function<bool(const SufficientChangePredicateCandidate &)>;
+
+    /**
+     * @brief Factory to generate a functor for "attribute was last reported" at least `minimumDurationMillis` ago.
+     *
+     * @param minimumDurationMillis - number of millis needed since last marked as dirty before we mark dirty again.
+     * @return a functor usable for the `changedPredicate` arg of `SetValue()`
+     */
+    static SufficientChangePredicate
+    GetPredicateForSufficientTimeSinceLastDirty(System::Clock::Milliseconds64 minimumDurationMillis)
+    {
+        return [minimumDurationMillis](const SufficientChangePredicateCandidate & candidate) -> bool {
+            return (candidate.lastDirtyValue != candidate.newValue) &&
+                ((candidate.nowTimestamp - candidate.lastDirtyTimestamp) >= minimumDurationMillis);
+        };
+    }
+
+    Nullable<T> value() const { return mValue; }
+    QuieterReportingPolicyFlags & policy() { return mPolicyFlags; }
+    const QuieterReportingPolicyFlags & policy() const { return mPolicyFlags; }
+
+    /**
+     * When this returns true, attribute should be marked for reporting. Auto-resets to false after call.
+     */
+    bool WasJustMarkedDirty()
+    {
+        bool wasDirty = mIsDirty;
+        mIsDirty      = false;
+        return wasDirty;
+    }
+
+    /**
+     * Force marking this attribute as dirty, with the `now` timestamp given as the reference point.
+     *
+     * WARNING: Only call `ForceDirty` after a `SetValue` call.
+     */
+    void ForceDirty(Timestamp now)
+    {
+        mIsDirty                  = true;
+        mLastDirtyTimestampMillis = now;
+        mLastDirtyValue           = mValue;
+    }
+
+    /**
+     * Set the updated value of the attribute, computing whether it needs to be reported according to `changedPredicate` and
+     * policies.
+     *
+     * - Any change of nullability between `newValue` and the old value will be considered dirty.
+     * - The policies from `QuieterReportingPolicyEnum` and set via `SetPolicy()` are self-explanatory by name.
+     * - The changedPredicate will be called with last dirty <timestamp, value> and new <timestamp value> and may override
+     *   the dirty state altogether when it returns true. Use sparingly and default to a functor returning false.
+     *
+     * Internal recording will be done about last dirty value and last dirty timestamp based on the policies having applied.
+     *
+     * @param newValue - new value to set for the attribute
+     * @param now - system monotonic timestamp at the time of the call
+     * @param changedPredicate - functor to possibly override dirty state
+     */
+    void SetValue(const chip::app::DataModel::Nullable<T> & newValue, Timestamp now, SufficientChangePredicate changedPredicate)
+    {
+        bool isChangeOfNull       = newValue.IsNull() ^ mValue.IsNull();
```

Comment:
any chance we could have a:

```cpp
if (mIsDirty) 
{
    mValue = newValue;
    return;
}
```

here to bypass all the computations below if already dirty? Then we can rely on mIsDirty being false below.

Or are we required to call changedPredicate on every newValue even if dirty?